clippy_lints/src/needless_borrows_for_generic_args.rs

```diff
@@ -230,9 +237,9 @@ fn needless_borrow_count<'tcx>(
 
         let referent_ty = cx.typeck_results().expr_ty(referent);
 
-        if (!is_copy(cx, referent_ty) && !referent_ty.is_ref())
-            && let ExprKind::AddrOf(_, _, inner) = reference.kind
-            && !matches!(inner.kind, ExprKind::Call(..) | ExprKind::MethodCall(..))
+        if !(is_copy(cx, referent_ty)
+            || referent_ty.is_ref() && referent_used_exactly_once(cx, possible_borrowers, reference)
```

Comment:
In the PR description you mention:

> To avoid emitting false suggestions, this PR reintroduces the `referent_used_exactly_once` logic and runs that check for referents that are themselves mutable references.

Where is the mutability checked?

---

Alos, when is this check ever run? `&T`/`&mut T` should both be copy, which should short circit the `||` on this line. :thinking:

Reply:
`&mut T` is *not* copy, as it must be unique. Observe how `#[derive(Clone, Copy)] struct Meow<'a>(&'a mut i32);` doesn't compile.

In most contexts it does behave like it is `Copy`, but under the hood it's actually a reborrow – and when passing a `&mut` to a function that takes `impl Trait`, the coercion that causes the reborrow doesn't happen.

Reply:
Ahh, interesting TIL. Then this checks out